toplevel/coqinit.ml

```diff
@@ -80,9 +80,10 @@ let toplevel_init_load_path () =
   (* first, developer specific directory to open *)
   ml_path_if Coq_config.local [coqlib/"dev"] @
 
-  (* main loops *)
-  ml_path_if (Coq_config.local || !Flags.boot) [coqlib/"stm"; coqlib/"ide"] @
-  ml_path_if (System.exists_dir (coqlib/"toploop")) [coqlib/"toploop"]
+  (* main loops, must uniformize topide directory *)
+  ml_path_if (Coq_config.local || !Flags.boot) [coqlib/"ide"] @
+  ml_path_if (System.exists_dir (coqlib/"topstm")) [coqlib/"topstm"] @
+  ml_path_if (System.exists_dir (coqlib/"topide")) [coqlib/"topide"]
```

Comment:
I don't see a semantic difference between "topstm" and "topide".
I think it is much simpler to have a single "toploop" directory where all toploops are installed.
Then, the code above, was just a hack to be able to load coqloops from a `-local` (not installed) coq setup.

Reply:
I think there is a very important difference as conceptually `ide` and `toplevel` are separate modules that could be built and installed separately, not using the hack we have now for CoqIDE.

Thus, it is better that each package install to their own directory so I would like to keep these two dirs for now.

Reply:
I disagree.

When coqdoc will interact with Coq directly, by providing a new toplevel plugin, what do we do? Add a `coqlib/"topdoc"` directory?

There is 1 kind of toplevel plugin (1 API, 1 loader), and there should be 1 directory where toplevel plugins should be installed. If only to avoid asking ourselves the question "which directory has to be scanned first when one passes `-toploop foo`).
The fact that, when Coq is not installed, we were scanning `ide/` and `stm/` was a hack to avoid creating a `toploop/` directoy at the root of Coq and copy there the cmxs.

Reply:
> When coqdoc will interact with Coq directly, by providing a new toplevel plugin, what do we do? Add a coqlib/"topdoc" directory?

If `coqdoc` is going to be a separate package as for the IDE, then it won't be `$coqlib/topdoc` but `$coqdoclib/toplevel` or whatever the thing is called. IMHO other packages have no business messing with `$coqlib`. [Apart from all the problems we get having to support a `local` flag, we'd rather get rid of it].

Reply:
I don't get it, why do you propose `topide/` then? if coqdoc it like the IDE then it should use `toploop/`.

Reply:
Well, I'd say let's address that when this futuristic `coqdoc` appears. For now, I still prefer this model as in fact I'd like to refactor CoqIDE file organization as to make it more componentized and in particular to allow it build over an already installed Coq.

Reply:
This has nothing to do with the new `topstm` and `topide` directories.
The `coqdoc` example is there just to show that the approach you propose makes little sense.
All toplevels should install their plugins in the `toplevel` direcotry, why should it be more complex than that?!

Reply:
And what should local builds do then? How is CoqIDE going to work if built locally vs a system-wide Coq?

Reply:
I think that you want 1 mllib file per directory, is it the case?

Reply:
No, you already have 2 in topstm.
Then please explain what you are trying to do here. Why can't these files stay where they are, why do we need two new directories.

Reply:
> I think that you want 1 mllib file per directory, is it the case?

This doesn't answer my question.



> No, you already have 2 in topstm.

and in `master` you have 4 `mllib` files in `stm`, and what do you mean?



> Then please explain what you are trying to do here. Why can't these files stay where they are, why do we need two new directories.

It is explained in the commit log, the workers depend on toplevel and toplevel in the stm so they shouldn't be there as they don't belong to the stm but to toplevel [which by the way the whole toplevel plugin system is a weird system but that's a diffferent story].

So indeed things that depend on `toplevel` are moved to a place above of toplevel, which is the correct way to do it in the current code (non) organization.

Having 1 or 2 directories is not relevant, that boils down to the IDE being a different component thus living on its own namespace. In fact, this commit is incomplete as the `ide` folder should indeed be organized as in `https://github.com/ejgallego/coqide-exp`.

Reply:
then put the stm mllibs in toplevel and leave the coqide one where it is (coqide depends on toplevel)